bentoml/configuration/containers.py

```diff
@@ -0,0 +1,172 @@
+                    " conform to the required schema."
+                ) from e
+
+        # Legacy configuration compatibility
+        if legacy_compatibility:
+            try:
+                self.config["api_server"]["port"] = config("apiserver").getint(
+                    "default_port"
+                )
+                self.config["api_server"]["max_request_size"] = config(
+                    "apiserver"
+                ).getint("default_max_request_size")
+                self.config["marshal_server"]["max_batch_size"] = config(
+                    "marshal_server"
+                ).getint("default_max_batch_size")
+                self.config["marshal_server"]["max_latency"] = config(
+                    "marshal_server"
+                ).getint("default_max_latency")
+                self.config["marshal_server"]["request_header_flag"] = config(
+                    "marshal_server"
+                ).get("marshal_request_header_flag")
+                self.config["yatai"]["url"] = config("yatai_service").get("url")
+                self.config["tracing"]["zipkin_api_url"] = config("tracing").get(
+                    "zipkin_api_url"
+                )
+                self.config["instrument"]["namespace"] = config("instrument").get(
+                    "default_namespace"
+                )
+            except KeyError as e:
+                raise BentoMLConfigException(
+                    "Overriding a non-existent configuration key in compatibility mode."
+                ) from e
+
+            if validate_schema:
+                try:
+                    SCHEMA.validate(self.config)
+                except SchemaError as e:
+                    raise BentoMLConfigException(
+                        "Configuration after applying legacy compatibility"
+                        " does not conform to the required schema."
+                    ) from e
+
+        # User override configuration
+        if override_config_file is not None and os.path.exists(override_config_file):
+            LOGGER.info("Applying user config override from %s" % override_config_file)
+            with open(override_config_file, "rb") as f:
+                override_config = YAML().load(f.read())
+            always_merger.merge(self.config, override_config)
+
+            if validate_schema:
+                try:
+                    SCHEMA.validate(self.config)
+                except SchemaError as e:
+                    raise BentoMLConfigException(
+                        "Configuration after user override does not conform to"
+                        " the required schema."
+                    ) from e
+
+    def override(self, keys: list, value):
+        if keys is None:
+            raise BentoMLConfigException("Configuration override key is None.")
+        if len(keys) == 0:
+            raise BentoMLConfigException("Configuration override key is empty.")
+        if value is None:
```

Comment:
Is it possible that the user is actually trying to set the value to None?

Reply:
This pattern will prevent the user from overriding a value to None. I don't know if we have any use cases where it would be valid for overriding the value to None. Maybe after we change the workers default to 1, the user might want BentoML to automatically set workers. In those use cases, we can perhaps use a special value, e.g. -1 or add another configuration flag `--automatic-workers` to allow this configuration.

Reply:
that makes sense, I was just curious if the user wants to reset a config value to the default, which can often be None. But I guess that's not supposed to happen because, with your new design, we encourage users to only have one source of config value. And I think an `auto worker` flag is a great idea, we should explore that.